Std/Data/String/Lemmas.lean

```diff
@@ -779,8 +826,66 @@ theorem mapAux_of_valid (f : Char → Char) : ∀ l r, mapAux f ⟨utf8Len l⟩
 theorem map_eq (f : Char → Char) (s) : map f s = ⟨s.1.map f⟩ := by
   simpa using mapAux_of_valid f [] s.1
 
--- TODO: substrEq
--- TODO: isPrefixOf
+/-! ### substrEq -/
+
+@[simp] theorem substrEq_loop_rfl (s : String) (off stp : Pos) : substrEq.loop s s off off stp := by
+  rw [substrEq.loop]
+  simp only [beq_self_eq_true, Bool.true_and, dite_eq_ite, ite_eq_right_iff]
+  intro h
+  have : stp.byteIdx - (off.byteIdx + csize (get s off)) < stp.byteIdx - off.byteIdx := by
+    rw [Nat.sub_add_eq]
+    exact Nat.sub_lt (Nat.sub_pos_of_lt h) (csize_pos _)
+  exact substrEq_loop_rfl s _ _
+termination_by stp.1 - off.1
+
+theorem substrEq_rfl (s : String) (off : Pos) (n : Nat) (h : off.byteIdx + n ≤ s.endPos.byteIdx) :
+  substrEq s off s off n := by
+  simp [substrEq, substrEq_loop_rfl]
+  exact h
+
+theorem substrEq_loop_self_append {s t : String} {off stp : Pos}
+  (h_off : off.Valid s) (h_stp : stp.Valid s) : substrEq.loop s (s ++ t) off off stp := by
+  rw [substrEq.loop]
+  simp only [dite_eq_ite, ite_eq_right_iff, Bool.and_eq_true, beq_iff_eq]
+  intro offb_lt_stpb
+  have off_lt_end := (Nat.lt_of_lt_of_le offb_lt_stpb h_stp.le_endPos)
+  have off_ne_end := (fun x => Nat.ne_of_lt off_lt_end <| Pos.ext_iff.mp x)
+  rw [get_append_of_valid h_off off_ne_end]
+  apply And.intro
+  · rfl
+  · have : stp.byteIdx - (off.byteIdx + csize (get s off)) < stp.byteIdx - off.byteIdx := by
+      simpa [Nat.sub_add_eq] using Nat.sub_lt (Nat.sub_pos_of_lt offb_lt_stpb) (csize_pos _)
+    exact substrEq_loop_self_append (valid_next h_off off_lt_end) h_stp
+termination_by stp.1 - off.1
+
+/-! ### isPrefixOf -/
+
+@[nolint unusedHavesSuffices, simp] -- false positive from unfolding substrEq.loop
```

Comment:
```suggestion
-- unusedHavesSuffices lint false positive from unfolding substrEq.loop
-- This will be fixed by https://github.com/leanprover/lean4/pull/4143
@[nolint unusedHavesSuffices, simp] 
```